backend/cmm_helpers.ml

```diff
@@ -1882,13 +1882,18 @@ let rec low_63 dbg e =
   | _ -> e
 
 (* CR-someday mshinwell/gbury: sign_extend_63 then tag_int should simplify to
-   just tag_int. Similarly, untag_int then sign_extend_63 should simplify to
-   untag_int. *)
+   just tag_int. *)
 let sign_extend_63 dbg e =
   check_64_bit_target "sign_extend_63";
-  let e = low_63 dbg e in
-  Cop
-    (Casr, [Cop (Clsl, [e; Cconst_int (1, dbg)], dbg); Cconst_int (1, dbg)], dbg)
+  match e with
+  | Cop (Casr, [_; Cconst_int (n, _)], _) when n > 0 ->
```

Comment:
There are architectures where n is seen modulo 64, so when n is 64, the previous shift would do nothing.
But this is kind of a programing error anyway since we consider that shifting by more than the word size is undefined in OCaml.

Reply:
Let's add this in a comment though please.

Reply:
I restricted the optimization to `n > 0 && n < 64` and fleshed out the comment.